hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/WALEntryStream.java

```diff
@@ -124,7 +124,7 @@ public enum HasNext {
    * You can call {@link #peek()} or {@link #next()} to get the actual {@link Entry} if this method
    * returns {@link HasNext#YES}.
    */
-  public HasNext hasNext() {
+  public HasNext hasNext() throws IOException {
```

Comment:
The design here is to not throwing any exceptions, instead, always use the return value to tell upper layers what to do...

Reply:
Hi, 
I have a small question here, are we seeing this issue only with meta replicas or user table region replicas as well?

Reply:
Then better wrap it with UncheckedIOException, or a specfic exception type defined by us own, add more comments to explain the reason, and then handle this exception in CatalogReplicationSourceWALReader?

Reply:
> Hi, I have a small question here, are we seeing this issue only with meta replicas or user table region replicas as well?

Only meta replicas. For user region replicas, we do use ZK stg for the wal queue.

Reply:
> Then better wrap it with UncheckedIOException, or a specfic exception type defined by us own, add more comments to explain the reason, and then handle this exception in CatalogReplicationSourceWALReader?

That's indeed better, makes the fix much simpler, as we don't even need additional CatalogReplicationSourceWALReader. Wrapping a FNFE in UncheckedIOException would go all the way up to the replication endpoint, which would then deal with it accordingly. Let me know WDYT, @Apache9 .